### PR TITLE
Pymoo model

### PR DESCRIPTION
**Pull request by samgdotson**
This PR adds a pymoo model class to instantiate a genetic algorithm problem as well as an "equations" module that allows users to construct objectives or constraints with the pymoo model. 

This closes issue #9

### CODE REVIEW

File: osier/equations.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import numpy as np
+
+class OsierEquation(ABC):
+
+    def __init__(self) -> None:
+        pass        
+
+    @abstractmethod
+    def _do(self, technology_list, X):
+        pass
```
Comment (yardasol) [inline osier/equations.py:11]: Consider adding a docstring to this function so its purpose is more clear.

File: osier/equations.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
+    """
+
+    dispatchable_techs = [t.technology_name for t in technology_list if t.dispatchable]
+
+    return dispatchable_techs
```
Comment (yardasol) [inline osier/equations.py]: It looks like you're omitting the `Return` docstring on these `get_*` functions. Does PEP8 allow this? If so, then the docstrings are fine as is as the functions themselves are very readable. If not, you should specify the type of variable returned by the `get_*` functions in their docstrings.

File: osier/equations.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+def get_tech_names(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
```
No such parameter exists for this function.

File: osier/equations.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+def get_dispatchable_techs(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
```
ditto

File: osier/equations.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+def get_nondispatchable_techs(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
```
ditto

File: osier/equations.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+def get_dispatchable_names(technology_list):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_results : :class:`pandas.DataFrame`
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
```
ditto

File: osier/equations.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+def annualized_capital_cost(technology_list, dispatch_model=None):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_model : :class:`osier.DispatchModel`
+        A _solved_ dispatch model (i.e. with model results and objective
+        attributes).
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
def annualized_capital_cost(technology_list):
    """
    Parameters
    ----------
    technology_list : list of :class:`osier.Technology` objects
        The list of technologies.
```

File: osier/equations.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+def annualized_fixed_cost(technology_list, dispatch_model=None):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_model : :class:`osier.DispatchModel`
+        A _solved_ dispatch model (i.e. with model results and objective
+        attributes).
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
def annualized_fixed_cost(technology_list):
    """
    Parameters
    ----------
    technology_list : list of :class:`osier.Technology` objects
        The list of technologies.
```

File: osier/equations.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    """
+    
+    dispatch_techs = get_dispatchable_techs(technology_list)
+    non_dispatch_techs = get_nondispatchable_techs(technology_list)
+    column_names = get_tech_names(technology_list)
+    dispatch_results = dispatch_model.results
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
def annual_co2(technology_list, solved_dispatch_model, emission='co2'):
    """
    Parameters
    ----------
    technology_list : list of :class:`osier.Technology` objects
        The list of technologies.
    solved_dispatch_model : :class:`osier.DispatchModel`
        A _solved_ dispatch model (i.e. with model results and objective
        attributes).
    """
    
    dispatch_techs = get_dispatchable_techs(technology_list)
    non_dispatch_techs = get_nondispatchable_techs(technology_list)
    column_names = get_tech_names(technology_list)
    dispatch_results = solved_dispatch_model.results
```

File: osier/equations.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+    Returns
+    -------
+    total_cost : float
+        The total system cost.
+    """
+
+    capital_cost = annualized_capital_cost(technology_list)
+    fixed_cost = annualized_fixed_cost(technology_list)
+    variable_cost = dispatch_model.objective
```
Comment (yardasol) [inline osier/equations.py]:
```suggestion
def total_cost(technology_list, solved_dispatch_model):
    """
    This function calculates the total system cost for a given 
    set of technologies and their corresponding dispatch.
    
    Parameters
    ----------
    technology_list : list of :class:`osier.Technology` objects
        The list of technologies.
    solved_dispatch_model : :class:`osier.DispatchModel`
        A _solved_ dispatch model (i.e. with model results and objective
        attributes).

    Returns
    -------
    total_cost : float
        The total system cost.
    """

    capital_cost = annualized_capital_cost(technology_list)
    fixed_cost = annualized_fixed_cost(technology_list)
    variable_cost = solved_dispatch_model.objective
```

File: tests/test_equations.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    hours = np.linspace(0, N, N)
+    demand = (np.sin((hours * np.pi / N_HOURS * 2 + phase_shift))
+              * -1 + np.ones(N) * (base_load + 1))
+
+    return demand
+
+
+def test_annualized_capital_cost(technology_set_1):
+    """
+    Tests the annualized cost is calculated correctly.
```
Comment (yardasol) [inline tests/test_equations.py]:
```suggestion
    Tests the annualized capital cost is calculated correctly.
```

File: tests/test_equations.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    expected = (nuclear.total_capital_cost / nuclear.lifetime) \
+               + (natural_gas.total_capital_cost / natural_gas.lifetime)
+    actual = annualized_capital_cost(technology_list=technology_set_1)
+
+    assert expected == pytest.approx(actual)
+
+
+def test_annualized_fixed_cost(technology_set_1):
+    """
+    Tests the annualized cost is calculated correctly.
```
Comment (yardasol) [inline tests/test_equations.py]:
```suggestion
    Tests the annualized fixed cost is calculated correctly.
```

File: tests/test_equations.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    expected = (nuclear.annual_fixed_cost) + (natural_gas.annual_fixed_cost)
+    actual = annualized_fixed_cost(technology_list=technology_set_1)
+
+    assert expected == pytest.approx(actual)
+
+
+def test_total_cost(technology_set_1, net_demand):
+    """
+    Tests that the :func:`total_cost` produces expected results. Where all
+    the technologies are simply :class:`Technology` objects.
```
Comment (yardasol) [inline tests/test_equations.py]:
```suggestion
    Tests that :func:`total_cost` produces expected results.
```

File: tests/test_equations.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+                + model.objective
+    actual = total_cost(technology_set_1, model)
+
+    assert expected == pytest.approx(actual)
+
+
+def test_annual_co2(technology_set_1, net_demand):
+    """
+    Tests that the :func:`annual_co2` produces expected results. Where all
+    the technologies are simply :class:`Technology` objects.
```
Comment (yardasol) [inline tests/test_equations.py]:
```suggestion
    Tests that :func:`annual_co2` produces expected results.
```

File: osier/models/capacity_expansion.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+import pandas as pd
+from copy import deepcopy
+import dill
+
+# Osier imports
+from osier import DispatchModel
+
+# Pymoo imports
+from pymoo.core.problem import Problem
+from pymoo.core.problem import ElementwiseProblem
```
Comment (yardasol) [inline osier/models/capacity_expansion.py]:
```suggestion
from pymoo.core.problem import Problem, ElementwiseProblem
```

File: osier/models/capacity_expansion.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
+# Standard imports
+import numpy as np
+import pandas as pd
+from copy import deepcopy
+import dill
+
+# Osier imports
+from osier import DispatchModel
+
+# Pymoo imports
```
Comment (yardasol) [inline osier/models/capacity_expansion.py]:
```suggestion
import numpy as np
import pandas as pd
from copy import deepcopy
import dill

from osier import DispatchModel

```

File: osier/models/capacity_expansion.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+
+
+LARGE_NUMBER = 1e40
+
+class CapacityExpansion(ElementwiseProblem):
+    """
+    The :class:`CapacityExpansion` class inherits from the 
+    :class:`pymoo.core.problem.ElementwiseProblem` class. This problem
+    determines the technology mix that _minimizes_ the provided
+    objectives.
```
Comment (yardasol) [inline osier/models/capacity_expansion.py:28]: Maybe we can massage this a little bit. It's not obvious to me what the `CapacityExpansion` class represents just from reading this docstring.

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @samgdotson, great work on this PR. Super clean stuff, I love to see it.

Most of my requested changes are typo fixes but there are also a few clarifying questions.

File: osier/equations.py (hunk 1 above)
Comment (samgdotson) [reply osier/equations.py:11]: This whole class is unused for the moment. I'm debating whether or not I want the objectives and constraints to be classes or functions. I went with classes for now since the implementation is much more straight forward. I could easily delete this.

File: osier/equations.py (hunk 2 above)
Comment (samgdotson) [reply osier/equations.py]: I'll specify the return variables in the doc string.

File: osier/equations.py (hunk 7 above)
Comment (samgdotson) [reply osier/equations.py]: The reason I have `dispatch_model=None` is due to the way I implemented objectives + constraints. The idea is to allow users to create their own objective or constraint functions, which get passed to the model class. There are only two things needed to calculate any quantity of interest -- the technology list and the dispatch results. However, not all objectives need both -- but I can't know that _a priori_ and it's simpler to just always pass both to the desired function rather than somehow check which pieces of data the function needs. (The metadata could about what info the function needs could be contained in a class like `OsierEquation` but I think that's still more difficult to implement and confusing to use). 

TL;DR The `dispatch_model` param is necessary.

File: osier/equations.py (hunk 8 above)
Comment (samgdotson) [reply osier/equations.py]: Ditto

File: osier/equations.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    return capital_cost
+
+
+def annualized_fixed_cost(technology_list, dispatch_model=None):
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_model : :class:`osier.DispatchModel`
```
Comment (samgdotson) [inline osier/equations.py]:
```suggestion
def annualized_fixed_cost(technology_list, solved_dispatch_model=None):
    """
    Parameters
    ----------
    technology_list : list of :class:`osier.Technology` objects
        The list of technologies.
    solved_dispatch_model : :class:`osier.DispatchModel`
```

File: osier/equations.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        The results dispatch results for each technology in 
+        :attr:`technology_list` from a :class:`osier.DispatchModel` run.
+    """
+
+    dispatchable_techs = [t.technology_name for t in technology_list if t.dispatchable]
+
+    return dispatchable_techs
+
+
+def annualized_capital_cost(technology_list, dispatch_model=None):
```
Comment (samgdotson) [inline osier/equations.py]:
```suggestion
def annualized_capital_cost(technology_list, solved_dispatch_model=None):
```

File: osier/equations.py [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+    """
+    Parameters
+    ----------
+    technology_list : list of :class:`osier.Technology` objects
+        The list of technologies.
+    dispatch_model : :class:`osier.DispatchModel`
```
Comment (samgdotson) [inline osier/equations.py]:
```suggestion
    solved_dispatch_model : :class:`osier.DispatchModel`
```

File: osier/equations.py (hunk 9 above)
Comment (samgdotson) [reply osier/equations.py]: Okay, I'll update the other functions to reflect this as well.

File: osier/models/capacity_expansion.py (hunk 17 above)
Comment (samgdotson) [reply osier/models/capacity_expansion.py:28]: Hmmm this class is just a `problem` that specifies how the decision space gets mapped onto objective space. This is not itself a genetic algorithm -- it's the map. This requires a bit of knowledge about how `pymoo` works but at this level I think that's a safe assumption (I'm slowly adding layers of abstraction and this layer is not the most abstract layer). I'm definitely open to ideas.

File: osier/equations.py (hunk 7 above)
Comment (yardasol) [reply osier/equations.py]: It sound like what you're saying is that you anticipate someone will extend this class and write their own `annualized_capital_cost` function that requires using the `dispatch_model` parameter, right?

File: osier/models/capacity_expansion.py (hunk 17 above)
Comment (yardasol) [reply osier/models/capacity_expansion.py:28]: The way I would do it is describe in general terms what `CapacityExpansion` is in terms of how it is used. Based on the name, perhaps something like the following would work: "This class represents a problem where the objective is to expand the capacity of technologys to a user-defined level". It's possible I'm misunderstanding this.

Comment (samgdotson) [reply osier/models/capacity_expansion.py:28]: The phrase "objective" has a pretty strict definition here and I don't want to confuse people. The "objective" for a user is to identify a solution set (obv) but the "objective(s)" of the problem are user-defined ("provided objectives"). I think the doc string is clear enough.

File: osier/equations.py (hunk 7 above)
Comment (samgdotson) [reply osier/equations.py]: A couple of things 
a) it's not a class -- just a function
b) users can write their own capital cost function, or use the one provided. 
c) In _general_ users can write any function they want that uses either a technology list, dispatch model, or both. These functions are passed as either objective arguments or constraint arguments to the `CapacityExpansion` class.
d) The `CapacityExpansion` class loops through the provided list of objectives and constraints, and assumes that they will accept a `technology_list` and a `dispatch_model`.

File: osier/models/capacity_expansion.py (hunk 17 above)
Comment (yardasol) [reply osier/models/capacity_expansion.py:28]: Okay. I think the confusion is just from my relative lack of familiarity with optimization terminology.

**Comment by yardasol** [comment]
Good work @samgdotson. Should I wait for another review or merge this now?

**Review by yardasol — APPROVED** [review]
Good job @samgdotson.
